components/fxa-client/src/pairing_channel/extension.rs

```diff
@@ -0,0 +1,555 @@
+        anyhow::bail!("Invalid extension");
+    }
+    fn get_type_tag(&self) -> u16 {
+        ExtensionType::PskKeyExchangeModes.into()
+    }
+
+    fn write(&self, message_type: HandshakeType, buf: &mut BytesMut) -> Result<()> {
+        match message_type {
+            HandshakeType::ClientHello => {
+                utils::write_with_callback_u8_len(buf, &mut |modes_buf| {
+                    self.modes.iter().for_each(|mode| {
+                        modes_buf.put_u8(*mode);
+                    });
+                    Ok(())
+                })?
+            }
+            _ => anyhow::bail!("Invalid handshake type"),
+        };
+        Ok(())
+    }
+
+    fn contains_version(&self, _: u16) -> bool {
+        false
+    }
+
+    fn is_selected_version(&self, _: u16) -> bool {
+        false
+    }
+}
+
+impl PskKeyExchangeExtension {
+    #[allow(dead_code)]
+    pub fn new(modes: Vec<u8>) -> Self {
+        Self { modes }
+    }
+    fn read(message_type: HandshakeType, buf: &mut Bytes) -> Result<Self> {
+        match message_type {
+            HandshakeType::ClientHello => {
+                let mut modes = Vec::new();
+                let modes_buf = utils::read_bytes_with_u8_len(buf)?;
+                let mut modes_buf = modes_buf.as_slice().to_bytes();
+                utils::read_callback_till_done(&mut modes_buf, &mut |buf| {
+                    modes.push(utils::read_u8(buf)?);
+                    Ok(())
+                })?;
+                Ok(Self { modes })
+            }
+            _ => anyhow::bail!("Invalid Handshake message"),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct UnrecognizedExtension {}
+
+impl ExtensionValue for UnrecognizedExtension {
+    fn get_binders(&self) -> Result<Vec<Vec<u8>>> {
+        anyhow::bail!("Invalid Ext")
+    }
+
+    fn get_identities(&self) -> Result<Vec<Vec<u8>>> {
+        anyhow::bail!("Invalid Ext")
+    }
+    fn get_modes(&self) -> Result<Vec<u8>> {
+        anyhow::bail!("Invalid ext");
+    }
+    fn get_selected_identity(&self) -> Result<u16> {
+        anyhow::bail!("Invalid extension");
+    }
+    fn get_type_tag(&self) -> u16 {
+        ExtensionType::UnspecifiedExtension.into()
+    }
+
+    fn write(&self, _: HandshakeType, _: &mut BytesMut) -> Result<()> {
+        Ok(())
+    }
+
+    fn contains_version(&self, _: u16) -> bool {
+        false
+    }
+
+    fn is_selected_version(&self, _: u16) -> bool {
+        false
+    }
+}
+
+impl UnrecognizedExtension {
+    fn new() -> Self {
+        Self {}
+    }
+}
+// Base struct for generic reading/writing of extensions,
+// which are all uniformly formatted as:
+//
+//   struct {
+//     ExtensionType extension_type;
+//     opaque extension_data<0..2^16-1>;
+//   } Extension;
+//
+// Extensions always appear inside of a handshake message,
+// and their internal structure may differ based on the
+// type of that message.
+#[derive(Debug)]
+pub struct Extension {
+    extension: Box<dyn ExtensionValue>,
```

Comment:
Oh, interesting! This seems similar to what I was suggesting for dealing with `ClientState`, but here I would (naively) have been tempted to go the other way and deliberately expose `Extension` as an enum, because calling code seems likely to want to match on the type of extension.